### ISSUE TITLE
Helm lint not pulling values outside of the default values file for schema validation

### ISSUE THREAD

**Issue by DavidOGrady** [helm/helm#8880]
Hi 

Im not 100% if this is a bug or the intended functionality of the helm v3 lint command but I noticed an issue when I uplifted our ci pipeline from helm v2 to helm v3 (see actual versions below)

What is currently happening?
when I run `helm lint . --values=second_values.yaml` any value that is in second_values.yaml that is also in the values.yaml file of the chart will not be overridden with the values of second_values.yaml. 
Example
![image](https://user-images.githubusercontent.com/22730434/95833505-a3f27080-0d33-11eb-90a1-fe040a46d2c7.png)

I was able to recreate the issue with a test chart (also attached as [testchart-0.1.0.zip](https://github.com/helm/helm/files/5369833/testchart-0.1.0.zip))

I dont know if this is a change in functionality. I am aware that the helm lint command handles schema checks differently between helm v2 and helm v3 however the schema used to validate the helm v3 lint command should be correct provided the second_values.yaml file is used.

Any help on this issue/question would be helpful as the only way we can work around this is by either disabling the helm lint command in our ci pipelines or removing some validation from our values.schema.json file

Output of `helm version`:
helm 3 installation
`helm version
version.BuildInfo{Version:"v3.3.4", GitCommit:"a61ce5633af99708171414353ed49547cf05013d", GitTreeState:"clean", GoVersion:"go1.14.9"}
`
helm 2 installation
`
helm2 version
Client: &version.Version{SemVer:"v2.16.5", GitCommit:"89bd14c1541fa93a09492010030fd3699ca65a97", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.16.1", GitCommit:"65cac393eeadea00384ea7108296ee6e5509c4de", GitTreeState:"clean"}
`

Output of `kubectl version`:
`kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.6", GitCommit:"96fac5cd13a5dc064f7d9f4f23030a6aeface6cc", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:49Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-16T17:47:13Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}`

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  N/A

**Comment by DavidOGrady** [comment]
I just want to note how the key pair values are declared in the values.yaml and the second_values.yaml file in the above chart. I believe the issue is more dialed in on the fact that if the value is defined but set to null in the values.yaml file, helm v3 lint will not over write it with the second_values.yaml key pair value

in values.yaml file
![image](https://user-images.githubusercontent.com/22730434/95836838-d2724a80-0d37-11eb-995c-d585ea71f792.png)

in second_values.yaml file (where the keys should be taken from but they dont seem to be in the lint command)
![image](https://user-images.githubusercontent.com/22730434/95836915-e7e77480-0d37-11eb-8a31-7eb1e9d10ecd.png)

**Comment by hickeyma** [comment]
@DavidOGrady Tried to reproduce it as follows:

```console
$ helm create issue-8880
Creating issue-8880

$ cat issue-8880/values.yaml

# Default values for issue-8880.
# This is a YAML-formatted file.
# Declare variables to be passed into your templates.

replicaCount: 1

[...]

service:
  type:
  port:

ingress:
[...]

$ helm lint issue-8880
==> Linting issue-8880
[INFO] Chart.yaml: icon is recommended
[ERROR] templates/: template: issue-8880/templates/NOTES.txt:8:39: executing "issue-8880/templates/NOTES.txt" at <.Values.service.type>: wrong type for value; expected string; got interface {}

Error: 1 chart(s) linted, 1 chart(s) failed

$ cat value-2.yaml 
service:
  type: ClusterIP
  port: 80

$ helm lint issue-8880 --values=value-2.yaml
==> Linting issue-8880
[INFO] Chart.yaml: icon is recommended

1 chart(s) linted, 0 chart(s) failed
```

Are you also using a [schema file](https://helm.sh/docs/topics/charts/#schema-files) with your chart? If so can you show it?

**Comment by DavidOGrady** [comment]
Hi @hickeyma 

Yes, I am using a schema file in the my test chart. You can see it in the attached zip file under the following path
testchart/values.schema.json

The zip can be seen here: 
[testchart-0.1.0.zip](https://github.com/helm/helm/files/5370436/testchart-0.1.0.zip)

This is what the schema looks like

{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "additionalProperties": true,
  "definitions": {
    "imagePullPolicy": {
      "type": "string",
      "enum": [
        "IfNotPresent",
        "Always",
        "Never"
      ]
    }
  },
  "properties": {
    "service": {
      "type": "object",
      "additionalProperties": true,
      "properties": {
        "port":{
          "type": "number"
        }
      },
      "required": [
        "port"
      ]
    }
  },
  "required": [
    "service"
  ]
}

**Comment by hickeyma** [comment]
Thanks for the info @DavidOGrady. I can reproduce it now as follows:

```console
$ cat issue-8880/values.schema.json
{
    "$schema": "http://json-schema.org/schema#",
    "type": "object",
    "properties": {
        "affinity": {
            "type": "object"
        },
[...]
"service": {
            "type": "object",
            "properties": {
                "port": {
                    "type": "integer"
                },
                "type": {
                    "type": "string"
                }
            },
            "required": [ "type", "port" ]
        },
[...]

$ helm lint issue-8880/
==> Linting issue-8880/
[INFO] Chart.yaml: icon is recommended
[ERROR] values.yaml: - service.port: Invalid type. Expected: integer, given: null
- service.type: Invalid type. Expected: string, given: null

[ERROR] templates/: values don't meet the specifications of the schema(s) in the following chart(s):
issue-8880:
- service: type is required
- service: port is required


Error: 1 chart(s) linted, 1 chart(s) failed

$ helm lint issue-8880 --values=value-2.yaml
==> Linting issue-8880
[INFO] Chart.yaml: icon is recommended
[ERROR] values.yaml: - service: type is required
- service: port is required


Error: 1 chart(s) linted, 1 chart(s) failed

$ helm lint issue-8880 --set service.port=80,service.type=ClusterIP
==> Linting issue-8880
[INFO] Chart.yaml: icon is recommended
[ERROR] values.yaml: - service: type is required
- service: port is required


Error: 1 chart(s) linted, 1 chart(s) failed
```

It does work for an install/upgrade:

```console
$ helm install issue-8880 issue-8880 --set service.port=80,service.type=ClusterIP --dry-run --debug
install.go:172: [debug] Original chart version: ""
install.go:189: [debug] CHART PATH: /home/marty/test/helm-charts/issue-8880

NAME: issue-8880
LAST DEPLOYED: Tue Oct 13 10:56:45 2020
NAMESPACE: default
STATUS: pending-install
REVISION: 1
USER-SUPPLIED VALUES:
service:
  port: 80
  type: ClusterIP
[...]
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=issue-8880,app.kubernetes.io/instance=issue-8880" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace default $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace default port-forward $POD_NAME 8080:$CONTAINER_PORT

$ helm install issue-8880 issue-8880 --values=value-2.yaml --dry-run --debug
install.go:172: [debug] Original chart version: ""
install.go:189: [debug] CHART PATH: /home/marty/test/helm-charts/issue-8880

NAME: issue-8880
LAST DEPLOYED: Tue Oct 13 10:58:09 2020
NAMESPACE: default
STATUS: pending-install
REVISION: 1
USER-SUPPLIED VALUES:
service:
  port: 80
  type: ClusterIP
[...]
```

**Comment by hickeyma** [comment]
This looks like a bug in linting. Going to set this to a bug and also update the title to include passing values using set also.

**Comment by DavidOGrady** [comment]
Thanks @hickeyma 
Yeah, I don't believe this is an issue in install upgrade or template command

**Comment by reitzig** [comment]
For reference, this appeared earlier as #7756. I'm still seeing it with Helm 3.4.0 and 3.5.2, and it's going to make us drop `helm lint` from our CI pipeline.

@donggangcj @hickeyma  @bacongobbler @jdolitsky @mluckam, is either of #8892 and #9260 close to being merged?

**Comment by hickeyma** [comment]
re-opened.

**Comment by mluckam** [comment]
This issue is addressed in #9260.  The code needs to be reviewed again after the request for comment was answered.

**Comment by abdennour** [comment]
`Invalid type. Expected: string, given: null`

**Comment by bacongobbler** [comment]
#9260 was only merged 6 days ago. It’ll be available in Helm 3.6.0. Until then feel free to test using the canary release.

**Comment by MrDeerly** [comment]
Is there any chance to reopen this issue? The bug seems to be back.
